packages/@lwc/engine-dom/src/styles.ts

```diff
@@ -0,0 +1,231 @@
+    }
+    contentsToStyleElements.set(content, elm);
+}
+
+function unsetStyleElementForTarget(target: ShadowRoot | Document, content: string) {
+    const contentsToStyleElements = targetsToStyleElements.get(target);
+    if (!isUndefined(contentsToStyleElements)) {
+        contentsToStyleElements.delete(content);
+    }
+}
+
+function getStyleElementForTarget(target: ShadowRoot | Document, content: string) {
+    const contentsToStyleElements = targetsToStyleElements.get(target);
+    if (!isUndefined(contentsToStyleElements)) {
+        return contentsToStyleElements.get(content);
+    }
+}
+
+function createOrGetConstructableStyleSheet(content: string) {
+    // It's important for CSSStyleSheets to be unique based on their content, so
+    // that adoptedStyleSheets.indexOf(sheet) works
+    let styleSheet = styleSheets.get(content);
+    if (isUndefined(styleSheet)) {
+        styleSheet = new CSSStyleSheet();
+        styleSheet.replaceSync(content);
+        styleSheets.set(content, styleSheet);
+    }
+    return styleSheet;
+}
+
+function getStylesToUsageCounts(target: ShadowRoot | Document) {
+    let stylesToCounts = stylesToUsageCount.get(target);
+    if (isUndefined(stylesToCounts)) {
+        stylesToCounts = new Map();
+        stylesToUsageCount.set(target, stylesToCounts);
+    }
+    return stylesToCounts;
+}
+
+function incrementOrDecrementUsageCount(
+    content: string,
+    target: ShadowRoot | Document,
+    delta: number
+) {
+    const stylesToCounts = getStylesToUsageCounts(target);
+
+    let count = stylesToCounts.get(content);
+    if (isUndefined(count)) {
+        count = 0;
+    }
+    count += delta;
+    stylesToCounts.set(content, count);
+    return count;
+}
+
+function insertConstructableStyleSheet(content: string, target: ShadowRoot | Document) {
+    const styleSheet = createOrGetConstructableStyleSheet(content);
+    const { adoptedStyleSheets } = target;
+    if (supportsMutableAdoptedStyleSheets) {
+        // This is only supported in later versions of Chromium:
+        // https://chromestatus.com/feature/5638996492288000
+        adoptedStyleSheets.push(styleSheet);
+    } else {
+        target.adoptedStyleSheets = [...adoptedStyleSheets, styleSheet];
+    }
+}
+
+function removeConstructableStyleSheet(content: string, target: ShadowRoot | Document) {
+    const styleSheet = createOrGetConstructableStyleSheet(content);
+    const { adoptedStyleSheets } = target;
+    if (supportsMutableAdoptedStyleSheets) {
+        // If indexOf() returns -1, then splice() won't do anything
+        adoptedStyleSheets.splice(adoptedStyleSheets.indexOf(styleSheet), 1);
+    } else {
+        target.adoptedStyleSheets = [...adoptedStyleSheets].filter((_) => _ !== styleSheet);
+    }
+}
+
+function insertStyleElement(content: string, target: ShadowRoot | Document) {
+    const elm = createStyleElement(content);
+    const targetAnchorPoint = isDocument(target) ? target.head : target;
```

Comment:
Previously we were doing:

```js
document.head || document.body || document
```

But I don't see the point of that. If it's the document, it has a head. And if it has neither a head nor a body, the browser won't let you append directly to the document anyway.